### DISCUSSION TITLE
Using trusted-replace-node-text

### DISCUSSION THREAD

**Discussion by peace2000** [uBlockOrigin/uAssets#23769] [Q&A]
Hi, can someone say how can I target this piece of text by using `trusted-replace-node-text`?

![image](https://github.com/uBlockOrigin/uAssets/assets/17256841/3917ee3b-1273-4801-b1a3-5fc0de43a55a)

I want to replace `120209` with `x`

I tried various filters but didn't work...

It's a private website so I cannot give a link nor account.

**Comment by stephenhawk8054** [comment]
Hmm... If it doesn't work, it might be because `rpnt` cannot inject in time.

I tried

```adb
web.archive.org##+js(rpnt, span, Save, ===================================)
```

for the page `https://web.archive.org/save` and it works. But if I try to do with some of the top `span` ones, it couldn't be injected in time.

**Reply by garry-ut99** [reply]
When I tried `*##+js(rpnt, SPAN, 120209, x)` in localhost page:
`<html><body><span class="text-name" data-bind="text: $data.name">120209</span></body></html>`




- in Firefox it didn't work for the first time, but then works afterwards
- in Chromium it works fine since the beginning

**Reply by peace2000** [reply]
I could get the filter to work on my page by adding `stay, 1` token. Otherwise it won't work. The same issue happens with FF and Chrome.

**Reply by garry-ut99** [reply]
There is also this token: `quitAfter, ms` for example: `quitAfter, 1000` or `quitAfter, 5000`, which will cause the scriptlet to stay only for some limited time instead of forever, if it works for your page and is sufficient for your case, it might be better than `stay, 1` because it won't potentially cause infinite overhead in case your page has dynamically updated DOM as mentioned https://github.com/gorhill/uBlock/wiki/Resources-Library#trusted-replace-node-textjs-.

**Reply by krystian3w** [reply]
If it doesn't have to work in Chrome(-mium) and you see these numbers in `view-source:`, I would check `$replace` (if the website is not encoded in an exotic standard or iso-8859-1).

```adb
|https://example.fi/foo?bar$doc,_,replace=/<span class="foo">33<\/span>/<span class="foo">x<\/span>/,_
```

**Reply by peace2000** [reply]
`quitAfter` doesn't work for this case, only `stay, 1` works. So I'm using it.

**Reply by peace2000** [reply]
`$replace` is not an option for this case as those texts don't exist in the page source code.

**Reply by stephenhawk8054** [reply]
@peace2000 Can you check other connections (XHR…)? If it doesn't appear in the source code, it has to be delivered from somewhere else.

**Reply by garry-ut99** [reply]
https://knockoutjs.com/documentation/text-binding.html
https://knockoutjs.com/examples/index.html
[publicwww.com: span class="text-name" data-bind="text: $data.name"](https://publicwww.com/websites/%3Cspan+class%3D%22text-name%22+data-bind%3D%22text%3A+%24data.name%22%3E/)

**Reply by peace2000** [reply]
@stephenhawk8054 I checked but the data doesn't appear in any requests I see in the logger.

**Comment by Yuki2718** [comment]
Is it possible to replace multiple places in a script at the same time? I want to merge
```
pks.raenonx.cc##+js(trusted-rpnt, script, /AdGuard.*?\\/, \)
pks.raenonx.cc##+js(trusted-rpnt, script, /Please.*?"/, \")
```
as using the both causes the site not to load. The target is this inline-script: 
```
self.__next_f.push([1,"Ldd\",null,{\"src\":\"/images/generic/globe.png\",\"alt\":\"About\",\"fill\":true,\"title\":\"About\",\"sizes\":\"(max-width: 768px) 10vw, 5vw\",\"className\":\"object-cover invert-hoverable\"}]}],[\"$\",\"div\",null,{\"className\":\"text-lg\",\"children\":\"About\"}]]}]\ndb:[\"$\",\"$Ldc\",null,{\"locale\":\"en\",\"localePrefix\":\"$undefined\",\"href\":\"https://discord.gg/eeveesleep\",\"prefetch\":false,\"className\":\"flex flex-row place-content-center place-items-center items-center text-center group button-clickable-glow size-12 shrink-0 self-end p-1\",\"target\":\"_blank\",\"children\":[\"$\",\"svg\",null,{\"xmlns\":\"http://www.w3.org/2000/svg\",\"viewBox\":\"0 0 127.14 96.36\",\"className\":\"invert-hoverable scale-75\",\"children\":[\"$\",\"path\",null,{\"fill\":\"#fff\",\"d\":\"M107.7,8.07A105.15,105.15,0,0,0,81.47,0a72.06,72.06,0,0,0-3.36,6.83A97.68,97.68,0,0,0,49,6.83,72.37,72.37,0,0,0,45.64,0,105.89,105.89,0,0,0,19.39,8.09C2.79,32.65-1.71,56.6.54,80.21h0A105.73,105.73,0,0,0,32.71,96.36,77.7,77.7,0,0,0,39.6,85.25a68.42,68.42,0,0,1-10.85-5.18c.91-.66,1.8-1.34,2.66-2a75.57,75.57,0,0,0,64.32,0c.87.71,1.76,1.39,2.66,2a68.68,68.68,0,0,1-10.87,5.19,77,77,0,0,0,6.89,11.1A105.25,105.25,0,0,0,126.6,80.22h0C129.24,52.84,122.09,29.11,107.7,8.07ZM42.45,65.69C36.18,65.69,31,60,31,53s5-12.74,11.43-12.74S54,46,53.89,53,48.84,65.69,42.45,65.69Zm42.24,0C78.41,65.69,73.25,60,73.25,53s5-12.74,11.44-12.74S96.23,46,96.12,53,91.08,65.69,84.69,65.69Z\"}]}]}]\n3d:[\"$\",\"$L35\",null,{\"locale\":\"en\",\"now\":\"$D2024-05-30T10:48:02.646Z\",\"timeZone\":\"UTC\",\"messages\":{\"UI\":{\"Subscription\":\"$37\",\"Root\":{\"#Comments\":[\"TODO: Add comments\"],\"Help\":\"Help / Documentations\"}}},\"children\":[\"$\",\"$Lde\",null,{\"locale\":\"en\"}]}]\n3e:[\"$\",\"$Ldf\",null,{\"announcements\":[{\"uuid\":\"f31ea303-726d-4cb6-8001-b5042389eb22\",\"message\":\"AdGuard needs to be disabled to make the website function properly, regardless ads-free or not.\",\"locale\":[\"en\"],\"level\":\"alert\",\"order\":30,\"expiry\":null},{\"uuid\":\"732e1024-3b06-4f69-b7bb-430c4684526b\",\"message\":\"Please do not report us to ad blockers. For details, please check [this doc](https://pks.raenonx.cc/en/docs/vi"])
```

**Reply by krystian3w** [reply]
Rather it will execute one after the other, it is left to poke around how to exclude so that one regex does not interfere with the other.

**Reply by peace2000** [reply]
@Yuki2718 

Would this do the trick?

`pks.raenonx.cc##+js(trusted-rpnt, script, /(?:AdGuard[a-z \.\,\-]*|Please[a-zA-Z \.\,\[\]\(:\/]*)/)`

**Reply by peace2000** [reply]
Hmm that blocks also (partially) another message....

**Reply by peace2000** [reply]
This rule should block the message without interfering other messages:

`pks.raenonx.cc##+js(trusted-rpnt, script, /(?:AdGuard[a-z \.\,\-]*|Please do not[a-zA-Z \.\,\[\]\(:\/]*)/)`

But doesn't work. "Please do not" message is not blocked but I think my rule should work as it works in regex101 also...

`https://regex101.com/r/9G9yQL/1`

**Reply by garry-ut99** [reply]
> Yuki2718: as using the both causes the site not to load. 

It's is not using them both at once which causes the breakage, it's the second one which does break the site, I get the breakage even if using only the second filter alone, it's because "Please..." is too wide as it removes the phrase in several different scripts... it should be changed to more strict "Please do not" to not break other scripts.

```
[pks.raenonx.cc][replace-node-text.fn ⁝ script ⁝ /AdGuard.*?\\/ ⁝ \] 
1 nodes present before installing mutation observer
[pks.raenonx.cc][replace-node-text.fn ⁝ script ⁝ /Please.*?"/ ⁝ \"] 
105 nodes present before installing mutation observer
```

> peace2000: But doesn't work. 

Your regex works partially and removes only the first message ("AdGuard...") but not the second message ("Please do not...") because your regex lacks a global modifer `g` ("Don't return after first match"), which you had enabled in `regex101.com`, but then you forgot to put it in your regex as well, that's why it finds & replaces only the first occurence and then stops and returns, your corrected filter will be: 
`pks.raenonx.cc##+js(trusted-rpnt, script, /(?:AdGuard[a-z \.\,\-]*|Please do not[a-zA-Z \.\,\[\]\(:\/]*)/g)`

By the way I also crafted my own filter, which uses "positive lookahead":
`pks.raenonx.cc##+js(trusted-rpnt, script, /(?:AdGuard needs|Please do not).*?(?=\\"|"])/g)`

Both of them work and provide the same result, so far I didn't spent time analyzing which one will be better.

Also as seen, I not only narrowed "Please" to "Please do not", but also "AdGuard" to "AdGuard needs" as well, to not risk possible breakages.

Also there seems to be some text leftovers due to a fact that the targetted message is splitted into two separate chunks / scripts, this version of filter will remove the remaining chunk as well:
`pks.raenonx.cc##+js(trusted-rpnt, script, /(?:AdGuard needs|Please do not|"ew\/site\/ads-and-credits\)\.).*?(?=\\"|"])/g)`. But there is a risk that: the leftover chunk might happen to belong to other message not targeted to removal, which will contain the same link, another thing is that it's not guaranteed that the leftover chunk will always look the same.

Optionally can be used "`condition, pattern`: A string or regex which must be found in the text content of the node in order for the substitution to occur" to safely replace only in the targeted script, and skip other scripts, as discussed in https://github.com/uBlockOrigin/uAssets/discussions/23855.

**Reply by Yuki2718** [reply]
> it's because "Please..." is too wide as it removes the phrase in several different scripts... it should be changed to more strict "Please do not" to not break other scripts.

Oh, I missed that. Thanks for pointing out.

**Comment by Yuki2718** [comment]
Is it possible to apply to items NOT including a specific text? I want to remove script with matching regex that does not include `PDF` for `exploader\.net`. IOW can `condition` be used negatively?

**Reply by gorhill** [reply]
I can add `excludes, pattern`. I will also rename `condition` to `includes` and deprecate `condition`, to be removed in some future.

**Reply by Yuki2718** [reply]
Will be the same for `rmnt`?

**Reply by gorhill** [reply]
Same helper function is used for `rpnt` and `rmnt`, so it should also work with `rmnt`.

**Reply by garry-ut99** [reply]
Exclusion/negation is possible with regex: 
- matches when doesn't contain a single word PDF : `condition, /^(?:(?!PDF).)*$/s`
- matches when doesn't contain any of words: PDF , word1 , word2: `condition, /^(?:(?!PDF|word1|word2).)*$/s`

https://sentry.io/answers/write-a-regular-expression-to-match-lines-not-containing-a-word/
https://stackoverflow.com/questions/406230/regular-expression-to-match-a-line-that-doesnt-contain-a-word

It works fine when excludiing simple text/words/regexes, 
however might be insufficient when trying to exclude some more complicated text/regex.

**Reply by garry-ut99** [reply]
`trusted-rpnt` doesn't work in [1.58.1rc8](https://github.com/gorhill/uBlock/releases/tag/1.58.1rc8)

**Reply by gorhill** [reply]
Fixed in 1.58.1rc9.

**Comment by Yuki2718** [comment]
NSFW `https://7mmtv.sx/ja/uncensored_content/20703/fc2-ppv-1237064.html` With `mmtv01.xyz##+js(trusted-rpnt, script, /openNewTab\(".*?"\)/g, null)` the logger says
> [mmtv01.xyz][replace-node-text.fn ⁝ script ⁝ /openNewTab\(".*?"\)/g ⁝ null] 1 nodes present before installing mutation observer

and fails to block popup on clicking the video. AG doesn't suffer, so maybe the scriptlet can be improved?

**Reply by stephenhawk8054** [reply]
Hmm... It replaces the text correctly for me (`null` appears)

![image](https://github.com/user-attachments/assets/7cc18d48-2479-4c19-9afd-ec3d59d01091)

even though I have the same `[mmtv01.xyz][replace-node-text.fn ⁝ script ⁝ /openNewTab(".*?")/g ⁝ null] 1 nodes present before installing mutation observer`

Can you check the devtools' script again?

**Reply by Yuki2718** [reply]
If fails on my Chrome

![Element](https://github.com/user-attachments/assets/d10e33a4-066e-457a-b44b-f5e448c67b68)

**Reply by Yuki2718** [reply]
> meanwhile can use to stop popup
> 
> ```
> mmtv01.xyz##+js(nowoif)
> mmtv01.xyz##+js(trusted-set-constant, playID, 1)
> ```

We already have nowoif but the second filter works.

**Reply by stephenhawk8054** [reply]
Hmm... Even when mine says `19 nodes present before installing mutation observer`, it still replaces the text normally

![image](https://github.com/user-attachments/assets/30e0322f-a16c-40ec-9826-546197d9f229)

**Reply by Yuki2718** [reply]
This seems to be a glitch in my Chrome profie (or uBO on it, but happened both on dev and stable). I noticed this earlier in https://github.com/uBlockOrigin/uBlock-discussions/discussions/900#discussioncomment-10116904 but thought it was my typo, but now even without typo it fails on the site too while works fine on another profile. IDK what is happening on this Chrome, nor what to look for.

**Reply by Yuki2718** [reply]
@ghajini I [added](https://github.com/uBlockOrigin/uAssets/commit/086fe9a3f7b9c7201cb79fd51c25d6be6d594a48) `trusted-rpnt` as it now works, but if not on your end feel free to change to `trusted-set-constant`.

**Reply by ghajini** [reply]
i don't know why **trusted-rpnt** does not work on my opera browser
- eg i load the page nsfw `https://7mmtv.sx/ja/uncensored_content/20703/fc2-ppv-1237064.html`
- click play => no popup
- now reload the page or click on player again(marked) ![Screenshot_20240724-151648-985.png](https://github.com/user-attachments/assets/78db74b8-0b6a-47b6-ab7d-2eb7897fd4ce)

- again popup reproduced
- even when you reload few times, popup comes back(you need to wait few seconds when click play)
- trusted-constant works for me consistently

**Reply by gorhill** [reply]
I cannot reproduce popups on my side, but a difference with AdGuard I see is that with AdGuard I do not have to click "Agree TOS", before being able to play while I do with uBO

**Reply by Yuki2718** [reply]
>  a difference with AdGuard I see is that with AdGuard I do not have to click "Agree TOS", before being able to play while I do with uBO

They have a `trusted-click-element` rule for that, but personally I don't feel it's necessary. Anyway, the problem seems to be something on my specific Chrome profile.

**Reply by krystian3w** [reply]
As it is added to AdGuard Cookie Notices, the reason is "laziness" or detecting damage to the site when another method* gets rid of Cookie Policy/GDPR.

*\* - like low premission scriptlets/redirects, `:remove()`, html filtering, CSS filters to unlock scrolling*

**Comment by peace2000** [comment]
Is it possible to use regex in the replacement part? Doesn't seem possible?

E.g. If I'd like to insert a linebreak, doesn't seem to work.

`github.com##+js(trusted-rpnt, #text, Suggest an answer, /\nSuggest an answer/)`

Results to:

![image](https://github.com/user-attachments/assets/5a125f11-8bab-4082-97b3-0e8bcf9f1792)

**Reply by gorhill** [reply]
Not possible.

**Reply by gorhill** [reply]
I could try to add support for this, but what is the actual case to justify adding code to handle this?
